ddspipe_participants/src/cpp/reader/dds/CommonReader.cpp

```diff
@@ -0,0 +1,264 @@
+        reckon_subscriber_qos_()
+    );
+
+    if (!dds_subscriber_)
+    {
+        throw utils::InitializationException(
+                  utils::Formatter() << "Error creating Subscriber for Participant " <<
+                      participant_id_ << " in topic " << topic_ << ".");
+    }
+
+    // Create CommonReader
+    // Listener must be set in creation as no callbacks should be missed
+    // It is safe to do so here as object is already created and callbacks do not require anything set in this method
+    reader_ = dds_subscriber_->create_datareader(
+        dds_topic_,
+        reckon_reader_qos_(),
+        nullptr);
+
+    if (!reader_)
+    {
+        throw utils::InitializationException(
+                  utils::Formatter() << "Error creating DataReader for Participant " <<
+                      participant_id_ << " in topic " << topic_ << ".");
+    }
+
+    // Set listener after entity creation to avoid SEGFAULT (produced when callback using reader_ is
+    // invoked before the variable is fully set)
+    reader_->set_listener(this);
+
+}
+
+void CommonReader::on_data_available(
+        fastdds::dds::DataReader* /* reader */)
+{
+    logInfo(DDSPIPE_DDS_READER, "On data available in reader in " << participant_id_ << " for topic " << topic_ << ".");
+    on_data_available_();
+}
+
+CommonReader::CommonReader(
+        const core::types::ParticipantId& participant_id,
+        const core::types::DdsTopic& topic,
+        const std::shared_ptr<core::PayloadPool>& payload_pool,
+        fastdds::dds::DomainParticipant* participant,
+        fastdds::dds::Topic* topic_entity)
+    : BaseReader(participant_id)
+    , dds_participant_(participant)
+    , dds_topic_(topic_entity)
+    , payload_pool_(payload_pool)
+    , topic_(topic)
+    , dds_subscriber_(nullptr)
+    , reader_(nullptr)
+{
+    // Do nothing
+}
+
+utils::ReturnCode CommonReader::take_nts_(
+        std::unique_ptr<core::IRoutingData>& data) noexcept
+{
+    // NOTE: we assume this function is called always from same thread
+    // NOTE: we assume this function is called always with nullptr data
+
+    logInfo(DDSPIPE_DDS_READER, "Taking data in " << participant_id_ << " for topic " << topic_ << ".");
+
+    // Check if there is data available
+    if (!(reader_->get_unread_count() > 0))
+    {
+        return utils::ReturnCode::RETCODE_NO_DATA;
+    }
+
+    RtpsPayloadData* rtps_data = new core::types::RtpsPayloadData();
+    data.reset(rtps_data);
+    fastdds::dds::SampleInfo info;
+
+    // Loop for read messages until we receive one that does not come from same participant
+    // NOTE: not reading local messages must be done in this loop because:
+    // 1. there is no way in DDS to ignore this reader from writer as there were in RTPS
+    // 2. ignore_local_endpoints would be override by xml configuration
+    while (true)
+    {
+        auto ret = reader_->take_next_sample(rtps_data, &info);
+
+        // If error reading data
+        if (!ret)
```

Comment:
The no data available is checked before `if (!(reader_->get_unread_count() > 0))`. And also the return code is transfered, so the user of this function could know if there was an error, and of which kind.

Reply:
Ok, I was confused by the loop. So after a second read I understood the problem about the endpoints in the same participant. Any solution in your mind to avoid the inefficiency? If this remains as it is, perhaps there should be a note in user documentation asking to always include this configuration in all user's xml profiles.